### ISSUE TITLE
Using facecolors='none' with a scatter plot results in off-centered symbols

### ISSUE THREAD

**Issue by stephjuneau** [matplotlib/matplotlib#17790]
### Bug report

**Bug summary**

Using `facecolors='none'` with matplotlib.pyplot.scatter(x,y) displays symbols that are offset from their input (x,y) positions. This is a different behavior compared to using, e.g., `facecolors='white'`, in which case the symbols are centered properly on the input (x,y) values.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.arange(8)
y = np.random.rand(8)

fig = plt.figure(figsize=(10,3))
plt.scatter(x,y,marker="o",s=80,facecolors='white',edgecolors='blue')
plt.scatter(x,y,s=2)
plt.scatter(x,y,marker="o",s=140,facecolors='none',edgecolors='red')
```

**Actual outcome**

```
<img width="611" alt="facecolors_bug" src="https://user-images.githubusercontent.com/26126576/85957121-93aeb500-b93f-11ea-9728-3fe8b7d7baa3.png">

```

**Expected outcome**

I would expect that the symbols should be properly centered and not offset when using `facecolors='none'` (red edge color in the example above). We can see on the figure that using `facecolors='white'` produces properly centered symbols (blue edge color). The red symbols should have the same centers.

**Matplotlib version**

  * Operating system: Scientific Linux release 6.10
  * Matplotlib version: 3.1.1
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.7.4
  * Jupyter version (if applicable): 4.3.0
  * Other libraries: numpy 1.17.2

 conda
 main default channel

**Comment by jklymak** [comment]
This is something to do with pixel snapping, though I don't understand why its different for "none" versus "white."  If you go to 600 dpi, the effect is un-noticable.  Thats not to excuse the problem, but maybe to point someone towards a solution....

**Comment by tacaswell** [comment]
I think the jitter is being caused by the different size, not be the different face colors:

```python
import numpy as np 
import matplotlib.pyplot as plt 
 
x = np.arange(8) 
y = np.random.rand(8) 
 
fig = plt.figure(figsize=(10,3)) 
plt.scatter(x,y,marker="o",s=80,facecolors='white',edgecolors='blue', alpha=.5)
plt.scatter(x,y,s=2) 
plt.scatter(x,y,marker="o",s=80,facecolors='none',edgecolors='red', alpha=.5)  
```

The issue is that to improve performance, if all of the markers are identical we render it once and then "stamp" it for each marker (which saves us from having to compute the anti-aliasing many times), The down side is that we can then miss the "correct" location of the marker by ~1/2 a pixel which should be "small", but when the figure is rendered at low resolution and then up-sampled on modern monitors the ~1 pixel offset in our rasterization can become a many screen-pixel offset that the user sees.

**Comment by timhoffm** [comment]
This is indeed related to the pixel snapping and stamping.

Interestingly, only the first two calls seem to result in snapping but the third call seems to render each marker individually.

**Comment by jklymak** [comment]
>  but when the figure is rendered at low resolution and then up-sampled on modern monitors the ~1 pixel offset in our rasterization can become a many screen-pixel offset that the user sees.

Oh, so I'm confused about what we do for hi-dpi monitors.  Do we render at the figure's dpi, and then upsample when we display?  I'd argue we should be moving towards doing the converse - hiDPI should be the default and then lowDPI downsampled...

**Comment by timhoffm** [comment]
I don't think this has to do with hi-dpi and up/downsampling. I can reproduce it with regular 100% display scaling.

**Comment by tacaswell** [comment]
The logic is https://github.com/matplotlib/matplotlib/blob/bdf01776977706975d98c6d61b1f345a062fb191/lib/matplotlib/collections.py#L374-L388 which picks which branch of https://github.com/matplotlib/matplotlib/blob/bdf01776977706975d98c6d61b1f345a062fb191/lib/matplotlib/collections.py#L396-L412 we go through (`draw_markers` is the stamping path).

**Comment by timhoffm** [comment]
Should we add a kwarg to prohibit single_path_optimization? For cases as the one above where we want to draw optimizable and non-optimizable markers together, this would allow for consistently positioning the markers.

**Comment by stephjuneau** [comment]
> I think the jitter is being caused by the different size, not be the different face colors:
> 
> ```python
> import numpy as np 
> import matplotlib.pyplot as plt 
>  
> x = np.arange(8) 
> y = np.random.rand(8) 
>  
> fig = plt.figure(figsize=(10,3)) 
> plt.scatter(x,y,marker="o",s=80,facecolors='white',edgecolors='blue', alpha=.5)
> plt.scatter(x,y,s=2) 
> plt.scatter(x,y,marker="o",s=80,facecolors='none',edgecolors='red', alpha=.5)  
> ```
> 
> The issue is that to improve performance, if all of the markers are identical we render it once and then "stamp" it for each marker (which saves us from having to compute the anti-aliasing many times), The down side is that we can then miss the "correct" location of the marker by ~1/2 a pixel which should be "small", but when the figure is rendered at low resolution and then up-sampled on modern monitors the ~1 pixel offset in our rasterization can become a many screen-pixel offset that the user sees.

When I try using the same size (s=80) for both, I can still see the offset between "white" and "none" in my case. It's less noticeable as the markers partially overlap but it's still there. I didn't try changing the resolution but I see other comments below that it may not be due to that so I'll hold off for now.

**Comment by stephjuneau** [comment]
Hi everyone, I just tested this bug again with Matplotlib version 3.7.1 and it still behaves as reported in the original issue. I tried using the same symbol size and everything else the same and just swapped `facecolors='white'` and `facecolors='none'` and only the open symbols are off centered.

**Comment by tacaswell** [comment]
Related to https://github.com/matplotlib/matplotlib/issues/26461

In one case we go through the "marker" fast-path and in the other we do not.

**Comment by dscontract** [comment]
Hello @stephjuneau,  

I’m following a **design-by-contract** approach, where the bug description acts as a requirement (what should *not* happen), and I generate a tentative patch that enforces those requirements.  

In this case, the issue came from using `facecolors='none'` in `plt.scatter`, which can trigger an **offset bug** in older versions of Matplotlib.  

### What changed:
- In the last `plt.scatter` call:
  - **Before:** `facecolors='none'` was used to make markers transparent.  
  - **After:** Replaced with `facecolors=(0, 0, 0, 0)` (an explicit fully transparent RGBA tuple).  
- This ensures transparency is preserved while avoiding the bug in older Matplotlib releases.  
- Added an inline comment explaining the fix.  

Could you try this patch and let me know if it resolves the issue on your side?  

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.arange(8)
y = np.random.rand(8)

fig = plt.figure(figsize=(10, 3))
plt.scatter(x, y, marker="o", s=80, facecolors='white', edgecolors='blue')
plt.scatter(x, y, s=2)
# Use a fully transparent facecolor to avoid the offset bug with facecolors='none' in older Matplotlib versions
plt.scatter(x, y, marker="o", s=140, facecolors=(0, 0, 0, 0), edgecolors='red')
```

**Comment by ayshih** [comment]
#32232 fixes this issue:
<img width="1000" height="300" alt="Image" src="https://github.com/user-attachments/assets/87bcf7e8-454f-4b3d-aba5-bd60ff3bda9a" />
